### DISCUSSION TITLE
Keyboard Interactive authentication in the upstream

### DISCUSSION THREAD

**Discussion by luisepbueno** [tg123/sshpiper#430] [Q&A]
Is there any way to handle keyboard interactive authentication in the upstream? Some Arista switches support only keyboard interactive auth. I need wait for the "password:" prompt and then send the password.

I would like to write a plugin with a function like this and set this function as the challenge for the upstream authentication. Is that possible?

`func (p *plugin) UpstreamKeyboardInteractiveChallenge(password []byte) (challenge *ssh.KeyboardInteractiveChallenge, err error) {
	var pwPromptRegex *regexp.Regexp
	if pwPromptRegex, err = regexp.Compile("[Pp]assword:"); err != nil {
		return
	}

	*challenge = func(name, instruction string, questions []string, echos []bool) (answers []string, err error) {
		for _, q := range questions {
			// Password prompt
			m := pwPromptRegex.Match([]byte(q))
			if m {
				answers = append(answers, string(password))
			}
		}

		return
	}

	return
}`

**Comment by tg123** [comment]
this is technically doable  
the api is not exposed to grpc based plugin.

but you can use https://github.com/tg123/sshpiper.crypto directly

you may need something like below

```
PiperConfig
{
xxxCallback -> Upstream {
   AuthMethod -> []{  ssh.KeyboardInteractive }
}
}
```
